operator/buffer/circular_file.go

```diff
@@ -0,0 +1,327 @@
+package buffer
+
+import (
+	"fmt"
+	"io"
+	"os"
+
+	"go.uber.org/multierr"
+)
+
+// circularFile is a io.ReadWriteCloser that writes to a fixed length file, such that
+// it wraps around to the beginning when reaching the end.
+// Methods on this struct are not thread-safe and require additional synchronization.
+type circularFile struct {
+	Start      int64
+	ReadPtr    int64
+	End        int64
+	Size       int64
+	Full       bool
+	f          fileLike
+	seekedRead bool
+	seekedEnd  bool
+	closed     bool
+	readPtrEnd bool
+}
+
+var _ io.ReadWriteCloser = (*circularFile)(nil)
+
+func openCircularFile(filePath string, sync bool, size int64) (*circularFile, error) {
+	fileFlags := os.O_CREATE | os.O_RDWR
+	if sync {
+		fileFlags |= os.O_SYNC
+	}
+
+	f, err := os.OpenFile(filePath, fileFlags, 0600)
+	if err != nil {
+		return nil, err
+	}
+
+	// Make sure that the file, if it already existed, is actually the correct size.
+	fsize, err := f.Seek(0, io.SeekEnd)
+	if err != nil {
+		fCloseErr := f.Close()
+		return nil, multierr.Combine(
+			err,
+			fCloseErr,
+		)
+	}
+
+	if fsize == 0 {
+		err := f.Truncate(size)
+		if err != nil {
+			fCloseErr := f.Close()
+			return nil, multierr.Combine(
+				err,
+				fCloseErr,
+			)
+		}
+	} else if fsize != size {
+		fCloseErr := f.Close()
+		return nil,
+			multierr.Combine(
+				fmt.Errorf("configured size (%d) does not match current on-disk size (%d)", size, fsize),
+				fCloseErr,
+			)
+	}
+
+	return &circularFile{
+		Size: size,
+		f:    f,
+	}, nil
+}
+
+func (rb *circularFile) Close() error {
+	if rb.closed {
+		return nil
+	}
+
+	rb.closed = true
+
+	err := rb.f.Close()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Read reads from the ring buffer into p, up to len(p) bytes.
+// The contents read are not discarded from the buffer; An independent read pointer
+// is maintained, which can be reset to the start of the buffer using ResetReadOffset()
+func (rb *circularFile) Read(p []byte) (int, error) {
+	if rb.closed {
```

Comment:
I think there might be a race condition here if Closed is called mid read. You call `Close` the file could be closed while reading leading to undefined behavior.

Reply:
circularFile itself is not thread safe, I handle that in the disk buffer itself because we're going to be locking there anyways, makes sense to just have this require additional sync, which I think just a normal os.File requires as well.